File: sdks/go/network.go
package cloudpool

import "context"

type NetworkClient struct {
	client *CloudPoolClient
}

type TunnelInfo struct {
	Active    bool   `json:"active"`
	LocalPort string `json:"localPort,omitempty"`
	RemoteURL string `json:"remoteUrl,omitempty"`
	URL       string `json:"url,omitempty"`
	ExpiresAt string `json:"expiresAt,omitempty"`
	StartedAt string `json:"startedAt,omitempty"`
}

type PubSubMessage struct {
	Channel string `json:"channel"`
	Message string `json:"message"`
}

type DomainInfo struct {
	Domain string `json:"domain"`
	Status string `json:"status"`
}

type GatewayUser struct {
	Email string `json:"email"`
	Role  string `json:"role"`
}

func (n *NetworkClient) TunnelStart(ctx context.Context, port, subdomain string) (*TunnelInfo, error) {
	var resp TunnelInfo
	err := n.client.PostJSON(ctx, "/api/network/tunnel/start", map[string]interface{}{
		"port": port, "subdomain": subdomain,
	}, nil, &resp)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

func (n *NetworkClient) TunnelStop(ctx context.Context) error {
	_, err := n.client.Post(ctx, "/api/network/tunnel/stop", nil, nil)
	return err
}

func (n *NetworkClient) TunnelStatus(ctx context.Context) (*TunnelInfo, error) {
	var resp TunnelInfo
	err := n.client.GetJSON(ctx, "/api/network/tunnel/status", nil, &resp)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

func (n *NetworkClient) PubSubBroadcast(ctx context.Context, channel, message string) (map[string]interface{}, error) {
	var resp map[string]interface{}
	err := n.client.PostJSON(ctx, "/api/pubsub/broadcast", PubSubMessage{
		Channel: channel, Message: message,
	}, nil, &resp)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

func (n *NetworkClient) ListDomains(ctx context.Context) ([]DomainInfo, error) {
	var resp []DomainInfo
	err := n.client.GetJSON(ctx, "/api/network/domains", nil, &resp)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

func (n *NetworkClient) GatewayRegister(ctx context.Context, email, password string) (map[string]interface{}, error) {
	var resp map[string]interface{}
	err := n.client.PostJSON(ctx, "/api/gateway-auth/register", map[string]string{
		"email": email, "password": password,
	}, nil, &resp)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

func (n *NetworkClient) GatewayList(ctx context.Context) ([]GatewayUser, error) {
	var resp []GatewayUser
	err := n.client.GetJSON(ctx, "/api/gateway-auth/list", nil, &resp)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

func (n *NetworkClient) GatewayDelete(ctx context.Context, email string) error {
	_, err := n.client.Delete(ctx, "/api/gateway-auth/"+email)
	return err
}

func (n *NetworkClient) Backups(ctx context.Context) (map[string]interface{}, error) {
	var resp map[string]interface{}
	err := n.client.GetJSON(ctx, "/api/gateway/backups", nil, &resp)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

File: sdks/python/cloudpool/services/network.py
"""Network service (tunnels, WAF, pubsub)."""

from __future__ import annotations

from typing import Any, Dict, List

from cloudpool._client import CloudPoolClient
from cloudpool.models.network import PubSubMessage, TunnelStatus, WafRule


class NetworkClient:
    """Synchronous network client.

    Manages tunnels, Web Application Firewall rules, and pub/sub
    messaging.

    Accessed via ``cloudpool.network`` on a ``CloudPool`` instance.
    """

    def __init__(self, client: CloudPoolClient) -> None:
        self._client = client

    # ── Tunnels ──

    def start_tunnel(self, port: int, subdomain: str = "") -> TunnelStatus:
        """Start a network tunnel.

        Args:
            port: Local port to tunnel.
            subdomain: Requested subdomain (optional).

        Returns:
            TunnelStatus with the public URL.
        """
        resp = self._client.request(
            "POST", "/api/v1/tunnels/start",
            json={"port": port, "subdomain": subdomain},
        )
        return TunnelStatus.from_dict(resp)

    def stop_tunnel(self, subdomain: str) -> None:
        """Stop a running tunnel.

        Args:
            subdomain: The tunnel's subdomain.
        """
        self._client.request("POST", f"/api/v1/tunnels/stop/{subdomain}")

    def get_tunnel_status(self, subdomain: str) -> TunnelStatus:
        """Get the status of a tunnel.

        Args:
            subdomain: The tunnel's subdomain.

        Returns:
            TunnelStatus with current state.
        """
        resp = self._client.request("GET", f"/api/v1/tunnels/status/{subdomain}")
        return TunnelStatus.from_dict(resp)

    # ── WAF ──

    def list_waf_rules(self, project_id: str) -> List[WafRule]:
        """List WAF rules for a project.

        Args:
            project_id: The project's unique identifier.

        Returns:
            List of WafRule objects.
        """
        resp = self._client.request("GET", f"/api/v1/projects/{project_id}/waf")
        return [WafRule.from_dict(r) for r in (resp if isinstance(resp, list) else [])]

    def add_waf_rule(self, project_id: str, rule_type: str, pattern: str, action: str) -> WafRule:
        """Add a WAF rule.

        Args:
            project_id: The project's unique identifier.
            rule_type: Rule type (e.g., "ip_blacklist", "rate_limit").
            pattern: Matching pattern.
            action: Action on match ("block", "allow", "challenge").

        Returns:
            The created WafRule.
        """
        resp = self._client.request(
            "POST", f"/api/v1/projects/{project_id}/waf",
            json={"ruleType": rule_type, "pattern": pattern, "action": action},
        )
        return WafRule.from_dict(resp)

    def delete_waf_rule(self, project_id: str, rule_id: str) -> None:
        """Delete a WAF rule.

        Args:
            project_id: The project's unique identifier.
            rule_id: The rule's unique identifier.
        """
        self._client.request("DELETE", f"/api/v1/projects/{project_id}/waf/{rule_id}")

    # ── Pub/Sub ──

    def broadcast(self, project_id: str, channel: str, payload: Any) -> PubSubMessage:
        """Broadcast a message to a PubSub channel.

        Args:
            project_id: The project's unique identifier.
            channel: Channel name.
            payload: Message payload (JSON-serializable).

        Returns:
            PubSubMessage with publication details.
        """
        resp = self._client.request(
            "POST", f"/api/v1/projects/{project_id}/pubsub/broadcast",
            json={"channel": channel, "payloadJson": payload},
        )
        return PubSubMessage.from_dict(resp)


class AsyncNetworkClient:
    """Asynchronous network client."""

    def __init__(self, client: Any) -> None:
        self._client = client

    async def start_tunnel(self, port: int, subdomain: str = "") -> TunnelStatus:
        resp = await self._client.request("POST", "/api/v1/tunnels/start", json={"port": port, "subdomain": subdomain})
        return TunnelStatus.from_dict(resp)

    async def stop_tunnel(self, subdomain: str) -> None:
        await self._client.request("POST", f"/api/v1/tunnels/stop/{subdomain}")

    async def get_tunnel_status(self, subdomain: str) -> TunnelStatus:
        resp = await self._client.request("GET", f"/api/v1/tunnels/status/{subdomain}")
        return TunnelStatus.from_dict(resp)

    async def list_waf_rules(self, project_id: str) -> List[WafRule]:
        resp = await self._client.request("GET", f"/api/v1/projects/{project_id}/waf")
        return [WafRule.from_dict(r) for r in (resp if isinstance(resp, list) else [])]

    async def add_waf_rule(self, project_id: str, rule_type: str, pattern: str, action: str) -> WafRule:
        resp = await self._client.request("POST", f"/api/v1/projects/{project_id}/waf", json={"ruleType": rule_type, "pattern": pattern, "action": action})
        return WafRule.from_dict(resp)

    async def delete_waf_rule(self, project_id: str, rule_id: str) -> None:
        await self._client.request("DELETE", f"/api/v1/projects/{project_id}/waf/{rule_id}")

    async def broadcast(self, project_id: str, channel: str, payload: Any) -> PubSubMessage:
        resp = await self._client.request("POST", f"/api/v1/projects/{project_id}/pubsub/broadcast", json={"channel": channel, "payloadJson": payload})
        return PubSubMessage.from_dict(resp)
